misc/multiaddr/src/onion_addr.rs

```diff
@@ -0,0 +1,82 @@
+use std::borrow::Cow;
+use std::fmt::Debug;
+use serde::export::Formatter;
+use arrayref::array_ref;
+use std::fmt;
+
+/// Represents an Onion v3 address
+#[derive(Clone)]
+pub struct Onion3Addr<'a>(Cow<'a, [u8; 35]>, u16);
+
+impl<'a> Onion3Addr<'a> {
+    /// Return the hash of the public key as bytes
+    pub fn hash(&self) -> &[u8; 35] {
+        self.0.as_ref()
+    }
+
+    /// Return the port
+    pub fn port(&self) -> u16 {
+        self.1
+    }
+
+    /// Consume this instance and create an owned version containing the same address
+    pub fn into_owned<'b>(self) -> Onion3Addr<'b> {
+        Self(Cow::Owned(self.0.into_owned()), self.1)
+    }
+}
+
+impl PartialEq for Onion3Addr<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        if self.1 != other.1 {
+            return false
+        }
+
+        for i in 0..35 {
+            if self.0[i] != other.0[i] {
+                return false
+            }
+        }
+
+        true
+    }
+}
+
+impl Eq for Onion3Addr<'_> { }
+
+impl From<([u8; 35], u16)> for Onion3Addr<'_> {
+    fn from(parts: ([u8; 35], u16)) -> Self {
+        Self(Cow::Owned(parts.0), parts.1)
+    }
+}
+
+impl<'a> From<(&'a [u8; 35], u16)> for Onion3Addr<'a> {
+    fn from(parts: (&'a [u8; 35], u16)) -> Self {
+        Self(Cow::Borrowed(parts.0), parts.1)
+    }
+}
+
+impl<'a> From<(&'a [u8], u16)> for Onion3Addr<'a> {
+    fn from(parts: (&'a [u8], u16)) -> Self {
+        Self(Cow::Borrowed(array_ref!(parts.0, 0, 35)), parts.1)
```

Comment:
I would not want to add this `From` impl as the slice may easily not be 35 bytes long which causes a panic. Maybe just remove it or make it a `TryFrom` if you really think it would be necessary.

Reply:
You are correct, they are not needed with your other suggestions in place - thanks